compiler/rustc_parse/src/parser/expr.rs

```diff
@@ -1522,17 +1522,18 @@ impl<'a> Parser<'a> {
                 Ok(this.mk_expr(this.prev_token.span, ExprKind::Underscore))
             } else if this.token.uninterpolated_span().at_least_rust_2018() {
                 // `Span::at_least_rust_2018()` is somewhat expensive; don't get it repeatedly.
-                if this.token.uninterpolated_span().at_least_rust_2024()
-                    // check for `gen {}` and `gen move {}`
-                    // or `async gen {}` and `async gen move {}`
-                    && (this.is_gen_block(kw::Gen, 0)
-                        || (this.check_keyword(kw::Async) && this.is_gen_block(kw::Gen, 1)))
-                {
-                    // FIXME: (async) gen closures aren't yet parsed.
-                    this.parse_gen_block()
-                } else if this.check_keyword(kw::Async) {
+                if this.check_keyword(kw::Async) {
                     // FIXME(gen_blocks): Parse `gen async` and suggest swap
-                    if this.is_gen_block(kw::Async, 0) {
+                    if this.is_gen_block(kw::Async, 0) || this.is_gen_block(kw::Gen, 1) {
+                        // Check for `async {` and `async move {`,
+                        this.parse_gen_block()
+                    } else {
+                        this.parse_expr_closure()
+                    }
+                } else if this.check_keyword(kw::Gen)
+                    && this.token.uninterpolated_span().at_least_rust_2024()
+                {
+                    if this.is_gen_block(kw::Gen, 0) {
```

Comment:
This deserves further cleanup, but I think it's already an improvement if all the `async` stuff always goes through the `kw::Async` arm.

Also we can now gate the edition check on the keyword check, yay